### PR TITLE
Add Josephus Problem

### PR DESCRIPTION
**Pull request by aqib-m31**
### Describe your change:
Added Josephus Problem.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by aqib-m31** [comment]
@tianyizheng02

File: maths/josephus_problem.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        step_size (int): Step size for elimination.
+
+    Returns:
+        int: The position of the last person remaining.
+
+    Examples:
+        >>> josephus_recursive(7, 3)
+        3
+        >>> josephus_recursive(10, 2)
+        4
```
Comment (cclauss) [inline maths/josephus_problem.py:33]:
```suggestion
        num_people: Number of people.
        step_size: Step size for elimination.

    Returns:
        The position of the last person remaining.

    Examples:
    >>> josephus_recursive(7, 3)
    3
    >>> josephus_recursive(10, 2)
    4
    >>> josephus_recursive(0, 2)
    ???
    >>> josephus_recursive(1.9, 2)
    ???
    >>> josephus_recursive(-2, 2)
    ???
    >>> josephus_recursive(7, 0)
    ???
    >>> josephus_recursive(7, -2)
    ???
    >>> josephus_recursive(1_000, 0.01)
    ???
```

Comment (aqib-m31) [reply maths/josephus_problem.py:33]: Could you kindly explain it to me?
@cclauss

Comment (aqib-m31) [reply maths/josephus_problem.py:33]: This program will break if these values are supplied, so I need to make a check before proceeding?

Comment (aqib-m31) [reply maths/josephus_problem.py:33]: >
Can I fix it like this?
```python
if (
        not isinstance(num_people, int)
        or not isinstance(step_size, int)
        or num_people <= 0
        or step_size <= 0
    ):
        raise ValueError("num_people or step_size is not a positive integer.")```

Comment (aqib-m31) [reply maths/josephus_problem.py:33]: @cclauss

**Comment by aqib-m31** [comment]
All set! @cclauss

**Review by cclauss — APPROVED** [review]
`josephus_recursive(996, 1)  # RecursionError on my machine`

`josephus_iterative(996, 1)  # Works just fine`

**Pull request merged**